File: 200-number-of-islands/number-of-islands.py
class Solution(object):
    def updateIsland(self, grid, i, j):
        # Base condition to stop recursion
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != '1':
            return

        # Mark the current cell as visited
        grid[i][j] = '0'

        # Explore all 4 directions
        self.updateIsland(grid, i - 1, j)  # up
        self.updateIsland(grid, i + 1, j)  # down
        self.updateIsland(grid, i, j - 1)  # left
        self.updateIsland(grid, i, j + 1)  # right

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0

        count = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    count += 1
                    self.updateIsland(grid, i, j)

        return count

File: 200-number-of-islands/number-of-islands.java
class Solution {
    
    private char[][] updateIland(char[][] grid,int i, int j ) {
        // update ij with 0
        grid[i][j] = '0';
        // check up down right left
        if( i-1 >= 0 && grid[i-1][j] == '1')
            grid = updateIland(grid, i-1, j);
        if( i+1 < grid.length && grid[i+1][j] == '1')
            grid = updateIland(grid, i+1, j);
        if( j-1 >= 0 && grid[i][j-1] == '1')
            grid = updateIland(grid, i, j-1);
        if( j+1 < grid[0].length && grid[i][j+1] == '1')
            grid = updateIland(grid, i, j+1);

        return grid;
    }

    public int numIslands(char[][] grid) {
        int count = 0;

        for(int i = 0; i < grid.length; i++) {
            for(int j = 0; j < grid[0].length; j++) {
                if(grid[i][j] == '1'){
                    count++;
                    grid = updateIland(grid,i,j);
                }
            }
        }

        return count;
    }
}